// @sendbird/uikit-react/cjs/OpenChannel/components/OpenChannelInput.js
'use strict';

var React = require('react');
var LocalizationContext = require('../../LocalizationContext-df436a16.js');
var ui_MessageInput = require('../../ui/MessageInput.js');
var OpenChannel_context = require('../../OpenChannelProvider-b431f910.js');
require('../../stringSet-d7f78de0.js');
require('../../index-0f42efcf.js');
require('../../_rollupPluginBabelHelpers-5fad415d.js');
require('prop-types');
require('../../const-0a12bf51.js');
require('../../const-692d0039.js');
require('../../ui/IconButton.js');
require('../../ui/Button.js');
require('../../index-c89616c9.js');
require('dompurify');
require('../../consts-2f667487.js');
require('../../ui/Icon.js');
require('../../index-33b67e46.js');
require('../../utils/message/getOutgoingMessageState.js');
require('../../consts-f0dc64d0.js');
require('../../consts-25fc4351.js');
require('../../tokenize-595877b4.js');
require('../../index-bf4e62fb.js');
require('../../UserProfileContext-0c255fa6.js');
require('../../compareIds-f064c47d.js');
require('../../topics-2431bb78.js');
require('@sendbird/chat');
require('@sendbird/chat/openChannel');
require('../../uuid-2953f4dd.js');
require('../../useSendbirdStateContext.js');
require('../../withSendbird.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

const MessageInputWrapper = (props, ref) => {
  const {
    currentOpenChannel,
    disabled,
    handleSendMessage,
    handleFileUpload,
    amIMuted
  } = OpenChannel_context.useOpenChannelContext();
  const channel = currentOpenChannel;
  const {
    stringSet
  } = React.useContext(LocalizationContext.LocalizationContext);
  const {
    value
  } = props;
  function getPlaceHolderString() {
    if (amIMuted) {
      return stringSet.MESSAGE_INPUT__PLACE_HOLDER__MUTED;
    }
    if (disabled) {
      return stringSet.MESSAGE_INPUT__PLACE_HOLDER__DISABLED;
    }
    return '';
  }
  if (!channel) {
    return null;
  }
  return /*#__PURE__*/React__default["default"].createElement("div", {
    className: "sendbird-openchannel-footer"
  }, /*#__PURE__*/React__default["default"].createElement(ui_MessageInput, {
    ref: ref,
    value: value,
    disabled: disabled,
    isVoiceMessageEnabled: false,
    onSendMessage: handleSendMessage,
    onFileUpload: handleFileUpload,
    placeholder: getPlaceHolderString()
  }));
};
var OpenChannelInput = /*#__PURE__*/React__default["default"].forwardRef(MessageInputWrapper);

module.exports = OpenChannelInput;
