// @sendbird/uikit-react/cjs/Thread/components/ThreadHeader.js
'use strict';

var React = require('react');
var ui_IconButton = require('../../ui/IconButton.js');
var ui_Icon = require('../../ui/Icon.js');
var ui_Label = require('../../index-c89616c9.js');
var ui_TextButton = require('../../ui/TextButton.js');
var LocalizationContext = require('../../LocalizationContext-df436a16.js');
require('../../_rollupPluginBabelHelpers-5fad415d.js');
require('prop-types');
require('../../stringSet-d7f78de0.js');
require('../../color-026d07a6.js');
require('../../index-0f42efcf.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function ThreadHeader(_ref) {
  let {
    className,
    channelName,
    renderActionIcon,
    onActionIconClick,
    onChannelNameClick
  } = _ref;
  const {
    stringSet
  } = LocalizationContext.useLocalization();
  const MemoizedActionIcon = React.useMemo(() => {
    if (typeof renderActionIcon === 'function') {
      return renderActionIcon({
        onActionIconClick
      });
    }
    return null;
  }, [renderActionIcon]);
  return /*#__PURE__*/React__default["default"].createElement("div", {
    className: `sendbird-thread-header ${className}`
  }, /*#__PURE__*/React__default["default"].createElement(ui_Label.Label, {
    className: "sendbird-thread-header__title",
    type: ui_Label.LabelTypography.H_2,
    color: ui_Label.LabelColors.ONBACKGROUND_1
  }, stringSet.THREAD__HEADER_TITLE), /*#__PURE__*/React__default["default"].createElement(ui_TextButton, {
    onClick: e => onChannelNameClick === null || onChannelNameClick === void 0 ? void 0 : onChannelNameClick(e),
    disableUnderline: true
  }, /*#__PURE__*/React__default["default"].createElement(ui_Label.Label, {
    className: "sendbird-thread-header__channel-name",
    type: ui_Label.LabelTypography.CAPTION_3,
    color: ui_Label.LabelColors.ONBACKGROUND_2
  }, channelName)), MemoizedActionIcon || /*#__PURE__*/React__default["default"].createElement("div", {
    className: "sendbird-thread-header__action"
  }, /*#__PURE__*/React__default["default"].createElement(ui_IconButton, {
    width: "32px",
    height: "32px",
    onClick: e => onActionIconClick(e)
  }, /*#__PURE__*/React__default["default"].createElement(ui_Icon["default"], {
    type: ui_Icon.IconTypes.CLOSE,
    fillColor: ui_Icon.IconColors.ON_BACKGROUND_1,
    width: "22px",
    height: "22px"
  }))));
}

module.exports = ThreadHeader;
