// @sendbird/uikit-react/cjs/Channel/components/RemoveMessageModal.js
'use strict';

var React = require('react');
var ui_Modal = require('../../ui/Modal.js');
var ui_Button = require('../../ui/Button.js');
var LocalizationContext = require('../../LocalizationContext-df436a16.js');
var Channel_context = require('../../ChannelProvider-da7b08a1.js');
require('react-dom');
require('../../index-dd9bc361.js');
require('../../ui/Icon.js');
require('prop-types');
require('../../ui/IconButton.js');
require('../../_rollupPluginBabelHelpers-5fad415d.js');
require('../../index-c89616c9.js');
require('../../stringSet-d7f78de0.js');
require('../../MediaQueryContext-f1e8fdbc.js');
require('../../index-0f42efcf.js');
require('../../UserProfileContext-0c255fa6.js');
require('../../useSendbirdStateContext.js');
require('../../withSendbird.js');
require('../../index-bf4e62fb.js');
require('../../topics-2431bb78.js');
require('../../index-33b67e46.js');
require('../../utils/message/getOutgoingMessageState.js');
require('../../compareIds-f064c47d.js');
require('../../const-692d0039.js');
require('@sendbird/chat/groupChannel');
require('../../uuid-2953f4dd.js');
require('@sendbird/chat/message');
require('../../consts-297fdae1.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

const RemoveMessage = props => {
  var _message$threadInfo;
  const {
    onSubmit = () => {/* noop */},
    onCancel,
    message
  } = props;
  const {
    stringSet
  } = React.useContext(LocalizationContext.LocalizationContext);
  const {
    deleteMessage
  } = Channel_context.useChannelContext();
  return /*#__PURE__*/React__default["default"].createElement(ui_Modal["default"], {
    type: ui_Button.ButtonTypes.DANGER,
    disabled: (message === null || message === void 0 ? void 0 : (_message$threadInfo = message.threadInfo) === null || _message$threadInfo === void 0 ? void 0 : _message$threadInfo.replyCount) > 0,
    onCancel: onCancel,
    onSubmit: () => {
      deleteMessage(message).then(() => {
        onSubmit();
        onCancel();
      });
    },
    submitText: stringSet.MESSAGE_MENU__DELETE,
    titleText: stringSet.MODAL__DELETE_MESSAGE__TITLE
  });
};

module.exports = RemoveMessage;
