// @sendbird/uikit-react/ChannelSettings/components/UserPanel.js
import React__default, { useContext, useState } from 'react';
import { a as LocalizationContext } from '../../LocalizationContext-076d6d2e.js';
import { L as Label, a as LabelTypography, b as LabelColors } from '../../index-63949de5.js';
import Icon, { IconTypes, IconColors } from '../../ui/Icon.js';
import Badge from '../../ui/Badge.js';
import { M as MemberList } from '../../MemberList-061d186f.js';
import { useChannelSettingsContext } from '../context.js';
import '../../stringSet-13bbd4a2.js';
import '../../index-99aabd9e.js';
import 'prop-types';
import '../../ui/Button.js';
import '../../ui/IconButton.js';
import '../../_rollupPluginBabelHelpers-6bb0305c.js';
import '../../ui/ContextMenu.js';
import 'react-dom';
import '../../ui/SortByRow.js';
import '../../uuid-1adedf62.js';
import '../../index-7eb65acf.js';
import '../../utils/message/getOutgoingMessageState.js';
import './UserListItem.js';
import '../../UserProfileContext-9cfbab4d.js';
import '../../ui/Avatar.js';
import '../../ui/ImageRenderer.js';
import '../../ui/MutedAvatarOverlay.js';
import '../../ui/UserProfile.js';
import '../../sendbirdSelectors.js';
import '../../topics-b2b77fb9.js';
import '../../utils-4bbbca2c.js';
import '../../useSendbirdStateContext.js';
import '../../withSendbird.js';
import '../../ui/Modal.js';
import '../../index-38e61c64.js';
import '../../MediaQueryContext-11f0e43f.js';
import '../../ui/UserListItem.js';
import '../../ui/Checkbox.js';

const kFormatter = num => {
  return Math.abs(num) > 999 ? `${(Math.abs(num) / 1000).toFixed(1)}K` : num;
};
const UserPanel = () => {
  const {
    stringSet
  } = useContext(LocalizationContext);
  const [showAccordion, setShowAccordion] = useState(false);
  const {
    channel
  } = useChannelSettingsContext();
  return /*#__PURE__*/React__default.createElement("div", {
    className: "sendbird-channel-settings__user-panel"
  }, /*#__PURE__*/React__default.createElement("div", {
    className: ['sendbird-channel-settings__panel-item', 'sendbird-channel-settings__members'].join(' '),
    role: "switch",
    "aria-checked": showAccordion,
    onKeyDown: () => setShowAccordion(!showAccordion),
    onClick: () => setShowAccordion(!showAccordion),
    tabIndex: 0
  }, /*#__PURE__*/React__default.createElement(Icon, {
    className: "sendbird-channel-settings__panel-icon-left",
    type: IconTypes.MEMBERS,
    fillColor: IconColors.PRIMARY,
    height: "24px",
    width: "24px"
  }), /*#__PURE__*/React__default.createElement(Label, {
    type: LabelTypography.SUBTITLE_1,
    color: LabelColors.ONBACKGROUND_1
  }, stringSet.CHANNEL_SETTING__MEMBERS__TITLE, /*#__PURE__*/React__default.createElement(Badge, {
    className: 'sendbird-channel-settings__badge',
    count: kFormatter(channel === null || channel === void 0 ? void 0 : channel.memberCount)
  })), /*#__PURE__*/React__default.createElement(Icon, {
    className: ['sendbird-channel-settings__panel-icon-right', 'sendbird-channel-settings__panel-icon--chevron', showAccordion ? 'sendbird-channel-settings__panel-icon--open' : ''].join(' '),
    type: IconTypes.CHEVRON_RIGHT,
    height: "24px",
    width: "24px"
  })), showAccordion && /*#__PURE__*/React__default.createElement(MemberList, null));
};

export { UserPanel as default };
