// @sendbird/uikit-react/ui/ReactionButton.js
import { b as _extends } from '../_rollupPluginBabelHelpers-6bb0305c.js';
import React__default from 'react';
import { u as useLongPress } from '../useLongPress-1303953d.js';
import { n as noop } from '../utils-4bbbca2c.js';
import '../MediaQueryContext-11f0e43f.js';

const ReactionButton = /*#__PURE__*/React__default.forwardRef((props, ref) => {
  const {
    className,
    width,
    height,
    selected,
    dataSbId = '',
    onClick,
    children
  } = props;
  const onClickHandler = useLongPress({
    onLongPress: noop,
    onClick
  }, {
    shouldPreventDefault: true,
    shouldStopPropagation: true
  });
  return /*#__PURE__*/React__default.createElement("div", _extends({
    className: [...(Array.isArray(className) ? className : [className]), `sendbird-reaction-button${selected ? '--selected' : ''}`].join(' '),
    ref: ref,
    role: "button",
    style: {
      width,
      height
    }
  }, onClickHandler, {
    tabIndex: 0,
    "data-sb-id": dataSbId
  }), /*#__PURE__*/React__default.createElement("div", {
    className: "sendbird-reaction-button__inner"
  }, children));
});

export { ReactionButton as default };
