// @sendbird/uikit-react/cjs/ChannelList.js
'use strict';

var React = require('react');
var ChannelList_context = require('./ChannelListProvider-6f1b6689.js');
var ChannelList_components_ChannelListUI = require('./ChannelList/components/ChannelListUI.js');
require('./_rollupPluginBabelHelpers-5fad415d.js');
require('@sendbird/chat/groupChannel');
require('./topics-2431bb78.js');
require('./uuid-2953f4dd.js');
require('./utils-279e7553.js');
require('./UserProfileContext-0c255fa6.js');
require('prop-types');
require('./useSendbirdStateContext.js');
require('./withSendbird.js');
require('./index-33b67e46.js');
require('./utils/message/getOutgoingMessageState.js');
require('./ChannelList/components/ChannelListHeader.js');
require('./LocalizationContext-df436a16.js');
require('./stringSet-d7f78de0.js');
require('./index-0f42efcf.js');
require('./index-c89616c9.js');
require('./ui/Avatar.js');
require('./ui/ImageRenderer.js');
require('./ui/Icon.js');
require('./ChannelList/components/AddChannel.js');
require('./ui/IconButton.js');
require('./CreateChannel.js');
require('./CreateChannel/components/CreateChannelUI.js');
require('./CreateChannelProvider-2f105a5c.js');
require('./sendbirdSelectors.js');
require('./CreateChannel/components/InviteUsers.js');
require('./ui/Modal.js');
require('react-dom');
require('./index-dd9bc361.js');
require('./ui/Button.js');
require('./MediaQueryContext-f1e8fdbc.js');
require('./ui/UserListItem.js');
require('./ui/MutedAvatarOverlay.js');
require('./ui/Checkbox.js');
require('./ui/UserProfile.js');
require('./ui/ContextMenu.js');
require('./ui/SortByRow.js');
require('./CreateChannel/components/SelectChannelType.js');
require('./ChannelList/components/ChannelPreview.js');
require('./ui/ChannelAvatar.js');
require('./utils-58e012fb.js');
require('./ui/Badge.js');
require('./index-4881e776.js');
require('./index-bf4e62fb.js');
require('./ui/Loader.js');
require('./index-21392a2e.js');
require('./index-80fc14fe.js');
require('./ui/MentionUserLabel.js');
require('./consts-2f667487.js');
require('./ui/TextButton.js');
require('./color-026d07a6.js');
require('./Channel/components/TypingIndicator.js');
require('./ChannelProvider-da7b08a1.js');
require('./compareIds-f064c47d.js');
require('./const-692d0039.js');
require('@sendbird/chat/message');
require('./consts-297fdae1.js');
require('./useLongPress-68ad7161.js');
require('./ChannelList/components/ChannelPreviewAction.js');
require('./EditUserProfile.js');
require('./index-606495e5.js');
require('./ui/Input.js');
require('./actionTypes-6cbf2054.js');
require('./index-162396f5.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

const ChannelList = props => {
  return /*#__PURE__*/React__default["default"].createElement(ChannelList_context.ChannelListProvider, {
    className: props === null || props === void 0 ? void 0 : props.className,
    disableUserProfile: props === null || props === void 0 ? void 0 : props.disableUserProfile,
    allowProfileEdit: props === null || props === void 0 ? void 0 : props.allowProfileEdit,
    onBeforeCreateChannel: props === null || props === void 0 ? void 0 : props.onBeforeCreateChannel,
    onThemeChange: props === null || props === void 0 ? void 0 : props.onThemeChange,
    overrideInviteUser: props === null || props === void 0 ? void 0 : props.overrideInviteUser,
    onProfileEditSuccess: props === null || props === void 0 ? void 0 : props.onProfileEditSuccess,
    onChannelSelect: props === null || props === void 0 ? void 0 : props.onChannelSelect,
    sortChannelList: props === null || props === void 0 ? void 0 : props.sortChannelList,
    queries: props === null || props === void 0 ? void 0 : props.queries,
    disableAutoSelect: props === null || props === void 0 ? void 0 : props.disableAutoSelect,
    activeChannelUrl: props === null || props === void 0 ? void 0 : props.activeChannelUrl,
    isTypingIndicatorEnabled: props === null || props === void 0 ? void 0 : props.isTypingIndicatorEnabled,
    isMessageReceiptStatusEnabled: props === null || props === void 0 ? void 0 : props.isMessageReceiptStatusEnabled
  }, /*#__PURE__*/React__default["default"].createElement(ChannelList_components_ChannelListUI, {
    renderChannelPreview: props === null || props === void 0 ? void 0 : props.renderChannelPreview,
    renderUserProfile: props === null || props === void 0 ? void 0 : props.renderUserProfile,
    renderHeader: props === null || props === void 0 ? void 0 : props.renderHeader,
    renderPlaceHolderEmptyList: props === null || props === void 0 ? void 0 : props.renderPlaceHolderEmptyList,
    renderPlaceHolderError: props === null || props === void 0 ? void 0 : props.renderPlaceHolderError,
    renderPlaceHolderLoading: props === null || props === void 0 ? void 0 : props.renderPlaceHolderLoading
  }));
};

module.exports = ChannelList;
