// @sendbird/uikit-react/cjs/Channel.js
'use strict';

var React = require('react');
var Channel_context = require('./ChannelProvider-da7b08a1.js');
var Channel_components_ChannelUI = require('./Channel/components/ChannelUI.js');
require('./UserProfileContext-0c255fa6.js');
require('prop-types');
require('./useSendbirdStateContext.js');
require('./withSendbird.js');
require('./_rollupPluginBabelHelpers-5fad415d.js');
require('./index-bf4e62fb.js');
require('./index-0f42efcf.js');
require('./topics-2431bb78.js');
require('./index-33b67e46.js');
require('./utils/message/getOutgoingMessageState.js');
require('./compareIds-f064c47d.js');
require('./const-692d0039.js');
require('@sendbird/chat/groupChannel');
require('./uuid-2953f4dd.js');
require('@sendbird/chat/message');
require('./consts-297fdae1.js');
require('./index-162396f5.js');
require('./LocalizationContext-df436a16.js');
require('./stringSet-d7f78de0.js');
require('./ui/Icon.js');
require('./index-c89616c9.js');
require('./ui/Loader.js');
require('./ui/ConnectionStatus.js');
require('./Channel/components/ChannelHeader.js');
require('./utils-080f7835.js');
require('./ui/IconButton.js');
require('./ui/ChannelAvatar.js');
require('./ui/Avatar.js');
require('./ui/ImageRenderer.js');
require('./utils-58e012fb.js');
require('./MediaQueryContext-f1e8fdbc.js');
require('./utils-279e7553.js');
require('./Channel/components/MessageList.js');
require('./Channel/components/Message.js');
require('./Channel/components/SuggestedMentionList.js');
require('./const-0a12bf51.js');
require('./ThreadProvider-654b9db0.js');
require('./Thread/context/types.js');
require('@sendbird/chat');
require('./ui/DateSeparator.js');
require('./color-026d07a6.js');
require('./ui/MessageInput.js');
require('./ui/Button.js');
require('dompurify');
require('./consts-2f667487.js');
require('./consts-f0dc64d0.js');
require('./consts-25fc4351.js');
require('./tokenize-595877b4.js');
require('./ui/MessageContent.js');
require('./ui/UserProfile.js');
require('./sendbirdSelectors.js');
require('./index-4881e776.js');
require('./index-21392a2e.js');
require('./index-80fc14fe.js');
require('./ui/MessageItemMenu.js');
require('./ui/ContextMenu.js');
require('react-dom');
require('./ui/SortByRow.js');
require('./types-ef51254c.js');
require('./ui/MessageItemReactionMenu.js');
require('./ui/ReactionButton.js');
require('./useLongPress-68ad7161.js');
require('./ui/EmojiReactions.js');
require('./ui/ReactionBadge.js');
require('./ui/BottomSheet.js');
require('./index-dd9bc361.js');
require('./ui/UserListItem.js');
require('./ui/MutedAvatarOverlay.js');
require('./ui/Checkbox.js');
require('./ui/Tooltip.js');
require('./ui/TooltipWrapper.js');
require('./Message/context.js');
require('./ui/AdminMessage.js');
require('./ui/TextMessageItemBody.js');
require('./index-450be53e.js');
require('./index.module-e5552393.js');
require('./ui/MentionLabel.js');
require('./ui/LinkLabel.js');
require('./ui/FileMessageItemBody.js');
require('./ui/TextButton.js');
require('./ui/ThumbnailMessageItemBody.js');
require('./ui/OGMessageItemBody.js');
require('./ui/UnknownMessageItemBody.js');
require('./ui/QuoteMessage.js');
require('./index-6e05ddef.js');
require('./ui/ThreadReplies.js');
require('./ui/VoiceMessageItemBody.js');
require('./ui/ProgressBar.js');
require('./VoicePlayer/useVoicePlayer.js');
require('./index-12ce8274.js');
require('./VoiceRecorder/context.js');
require('./ui/PlaybackTime.js');
require('./Channel/components/FileViewer.js');
require('./Channel/components/RemoveMessageModal.js');
require('./ui/Modal.js');
require('./index-a715eaef.js');
require('./useDirtyGetMentions-cc7de01f.js');
require('./Channel/components/UnreadCount.js');
require('./Channel/components/FrozenNotification.js');
require('./Channel/components/TypingIndicator.js');
require('./Channel/components/MessageInput.js');
require('./ui/QuoteMessageInput.js');
require('./VoiceMessageInputWrapper-2aaed45b.js');
require('./VoiceRecorder/useVoiceRecorder.js');
require('./index-9b38e167.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

const Channel = props => {
  return /*#__PURE__*/React__default["default"].createElement(Channel_context.ChannelProvider, {
    channelUrl: props === null || props === void 0 ? void 0 : props.channelUrl,
    isReactionEnabled: props === null || props === void 0 ? void 0 : props.isReactionEnabled,
    isMessageGroupingEnabled: props === null || props === void 0 ? void 0 : props.isMessageGroupingEnabled,
    showSearchIcon: props === null || props === void 0 ? void 0 : props.showSearchIcon,
    animatedMessage: props === null || props === void 0 ? void 0 : props.animatedMessage,
    highlightedMessage: props === null || props === void 0 ? void 0 : props.highlightedMessage,
    startingPoint: props === null || props === void 0 ? void 0 : props.startingPoint,
    onBeforeSendUserMessage: props === null || props === void 0 ? void 0 : props.onBeforeSendUserMessage,
    onBeforeSendFileMessage: props === null || props === void 0 ? void 0 : props.onBeforeSendFileMessage,
    onBeforeUpdateUserMessage: props === null || props === void 0 ? void 0 : props.onBeforeUpdateUserMessage,
    onChatHeaderActionClick: props === null || props === void 0 ? void 0 : props.onChatHeaderActionClick,
    onSearchClick: props === null || props === void 0 ? void 0 : props.onSearchClick,
    onBackClick: props === null || props === void 0 ? void 0 : props.onBackClick,
    replyType: props === null || props === void 0 ? void 0 : props.replyType,
    threadReplySelectType: props === null || props === void 0 ? void 0 : props.threadReplySelectType,
    queries: props === null || props === void 0 ? void 0 : props.queries,
    renderUserProfile: props === null || props === void 0 ? void 0 : props.renderUserProfile,
    filterMessageList: props === null || props === void 0 ? void 0 : props.filterMessageList,
    disableUserProfile: props === null || props === void 0 ? void 0 : props.disableUserProfile,
    disableMarkAsRead: props === null || props === void 0 ? void 0 : props.disableMarkAsRead,
    onReplyInThread: props === null || props === void 0 ? void 0 : props.onReplyInThread,
    onQuoteMessageClick: props === null || props === void 0 ? void 0 : props.onQuoteMessageClick,
    onMessageAnimated: props === null || props === void 0 ? void 0 : props.onMessageAnimated,
    onMessageHighlighted: props === null || props === void 0 ? void 0 : props.onMessageHighlighted
  }, /*#__PURE__*/React__default["default"].createElement(Channel_components_ChannelUI, {
    isLoading: props === null || props === void 0 ? void 0 : props.isLoading,
    renderPlaceholderLoader: props === null || props === void 0 ? void 0 : props.renderPlaceholderLoader,
    renderPlaceholderInvalid: props === null || props === void 0 ? void 0 : props.renderPlaceholderInvalid,
    renderPlaceholderEmpty: props === null || props === void 0 ? void 0 : props.renderPlaceholderEmpty,
    renderChannelHeader: props === null || props === void 0 ? void 0 : props.renderChannelHeader,
    renderMessage: props === null || props === void 0 ? void 0 : props.renderMessage,
    renderMessageInput: props === null || props === void 0 ? void 0 : props.renderMessageInput,
    renderTypingIndicator: props === null || props === void 0 ? void 0 : props.renderTypingIndicator,
    renderCustomSeparator: props === null || props === void 0 ? void 0 : props.renderCustomSeparator,
    renderFileUploadIcon: props === null || props === void 0 ? void 0 : props.renderFileUploadIcon,
    renderVoiceMessageIcon: props === null || props === void 0 ? void 0 : props.renderVoiceMessageIcon,
    renderSendMessageIcon: props === null || props === void 0 ? void 0 : props.renderSendMessageIcon
  }));
};

module.exports = Channel;
