// @sendbird/uikit-react/cjs/consts-297fdae1.js
'use strict';

const SCROLL_BUFFER = 10;

// voice message record
const VOICE_RECORDER_CLICK_BUFFER_TIME = 250;
const VOICE_RECORDER_DEFAULT_MIN = 1000; // 1 seconds
const VOICE_RECORDER_DEFAULT_MAX = 600000; // 10 minutes
const VOICE_RECORDER_AUDIO_BITS = 128000;
const BROWSER_SUPPORT_MIME_TYPE_LIST = ['audio/webm', 'audio/mp4', 'audio/mpeg', 'audio/ogg'];

/**
 * Append Audio element to the root
 * because I can't get the Audio element in the useEffect unmount scope
 */
const VOICE_PLAYER_ROOT_ID = 'sendbird-voice-player-provider-root';
const VOICE_PLAYER_AUDIO_ID = 'sendbird-global-audio-player-id';

// voice message file
const VOICE_MESSAGE_FILE_NAME = 'Voice_message.mp3';
const VOICE_MESSAGE_MIME_TYPE = 'audio/mp3;sbu_type=voice';

// meta array
const META_ARRAY_VOICE_DURATION_KEY = 'KEY_VOICE_MESSAGE_DURATION';
const META_ARRAY_MESSAGE_TYPE_KEY = 'KEY_INTERNAL_MESSAGE_TYPE';
const META_ARRAY_MESSAGE_TYPE_VALUE__VOICE = 'voice/mp3';

exports.BROWSER_SUPPORT_MIME_TYPE_LIST = BROWSER_SUPPORT_MIME_TYPE_LIST;
exports.META_ARRAY_MESSAGE_TYPE_KEY = META_ARRAY_MESSAGE_TYPE_KEY;
exports.META_ARRAY_MESSAGE_TYPE_VALUE__VOICE = META_ARRAY_MESSAGE_TYPE_VALUE__VOICE;
exports.META_ARRAY_VOICE_DURATION_KEY = META_ARRAY_VOICE_DURATION_KEY;
exports.SCROLL_BUFFER = SCROLL_BUFFER;
exports.VOICE_MESSAGE_FILE_NAME = VOICE_MESSAGE_FILE_NAME;
exports.VOICE_MESSAGE_MIME_TYPE = VOICE_MESSAGE_MIME_TYPE;
exports.VOICE_PLAYER_AUDIO_ID = VOICE_PLAYER_AUDIO_ID;
exports.VOICE_PLAYER_ROOT_ID = VOICE_PLAYER_ROOT_ID;
exports.VOICE_RECORDER_AUDIO_BITS = VOICE_RECORDER_AUDIO_BITS;
exports.VOICE_RECORDER_CLICK_BUFFER_TIME = VOICE_RECORDER_CLICK_BUFFER_TIME;
exports.VOICE_RECORDER_DEFAULT_MAX = VOICE_RECORDER_DEFAULT_MAX;
exports.VOICE_RECORDER_DEFAULT_MIN = VOICE_RECORDER_DEFAULT_MIN;
