// @sendbird/uikit-react/cjs/CreateOpenChannel.js
'use strict';

var React = require('react');
var CreateOpenChannel_components_CreateOpenChannelUI = require('./CreateOpenChannel/components/CreateOpenChannelUI.js');
var CreateOpenChannel_context = require('./CreateOpenChannel/context.js');
require('./LocalizationContext-df436a16.js');
require('./stringSet-d7f78de0.js');
require('./index-0f42efcf.js');
require('./ui/Avatar.js');
require('./ui/ImageRenderer.js');
require('./ui/Icon.js');
require('prop-types');
require('./uuid-2953f4dd.js');
require('./ui/Button.js');
require('./index-c89616c9.js');
require('./ui/Modal.js');
require('react-dom');
require('./index-dd9bc361.js');
require('./ui/IconButton.js');
require('./_rollupPluginBabelHelpers-5fad415d.js');
require('./MediaQueryContext-f1e8fdbc.js');
require('./ui/Input.js');
require('./ui/TextButton.js');
require('./color-026d07a6.js');
require('./useSendbirdStateContext.js');
require('./withSendbird.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function CreateOpenChannel(_ref) {
  let {
    className,
    onCreateChannel,
    onBeforeCreateChannel,
    closeModal,
    renderHeader,
    renderProfileInput
  } = _ref;
  return /*#__PURE__*/React__default["default"].createElement(CreateOpenChannel_context.CreateOpenChannelProvider, {
    className: className,
    onCreateChannel: onCreateChannel,
    onBeforeCreateChannel: onBeforeCreateChannel
  }, /*#__PURE__*/React__default["default"].createElement(CreateOpenChannel_components_CreateOpenChannelUI, {
    closeModal: closeModal,
    renderHeader: renderHeader,
    renderProfileInput: renderProfileInput
  }));
}

module.exports = CreateOpenChannel;
