// @sendbird/uikit-react/cjs/ui/ImageRenderer.js
'use strict';

var React = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

/*
  ImageRenderer displays image with url or source
  it checks if the source exist with img tag first
  if it exists onLoad is called, if not onError is called
  and those properties switch img tag to real purposing element
*/

const ImageRenderer = _ref => {
  let {
    className = '',
    url,
    alt = '',
    width = null,
    height = null,
    circle = false,
    fixedSize = false,
    placeHolder = null,
    defaultComponent = null,
    onLoad = () => {/* noop */},
    onError = () => {/* noop */}
  } = _ref;
  const [showDefaultComponent, setShowDefaultComponent] = React.useState(false);
  const [showPlaceHolder, setShowPlaceHolder] = React.useState(true);
  const DefaultComponent = React.useMemo(() => {
    return typeof defaultComponent === 'function' ? defaultComponent() : defaultComponent;
  }, [defaultComponent]);
  const PlaceHolder = React.useMemo(() => {
    return placeHolder && typeof placeHolder === 'function' ? placeHolder({
      style: {
        width: '100%',
        minWidth: width,
        maxWidth: fixedSize ? width : '400px',
        height,
        position: 'absolute',
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center'
      }
    }) : null;
  }, [placeHolder]);
  const HiddenImageLoader = React.useMemo(() => {
    setShowDefaultComponent(false);
    // reset the state when url is changed
    return /*#__PURE__*/React__default["default"].createElement("img", {
      className: "sendbird-image-renderer__hidden-image-loader",
      src: url,
      alt: alt,
      onLoad: () => {
        setShowPlaceHolder(false);
        onLoad();
      },
      onError: () => {
        setShowDefaultComponent(true);
        onError();
      }
    });
  }, [url]);
  return /*#__PURE__*/React__default["default"].createElement("div", {
    className: [...(Array.isArray(className) ? className : [className]), 'sendbird-image-renderer'].join(' '),
    style: {
      width: '100%',
      minWidth: width,
      maxWidth: fixedSize ? width : '400px',
      height
    }
  }, showPlaceHolder && PlaceHolder, showDefaultComponent ? DefaultComponent : /*#__PURE__*/React__default["default"].createElement("div", {
    className: "sendbird-image-renderer__image",
    style: {
      width: '100%',
      minWidth: width,
      maxWidth: fixedSize ? width : '400px',
      height,
      position: 'absolute',
      backgroundRepeat: 'no-repeat',
      backgroundPosition: 'center',
      backgroundSize: 'cover',
      backgroundImage: `url(${url})`,
      borderRadius: circle ? '50%' : null
    }
  }), HiddenImageLoader);
};

module.exports = ImageRenderer;
