// @sendbird/uikit-react/cjs/ui/MutedAvatarOverlay.js
'use strict';

var React = require('react');
var ui_Icon = require('./Icon.js');
require('prop-types');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function MutedAvatarOverlay(props) {
  const {
    height = 24,
    width = 24
  } = props;
  return /*#__PURE__*/React__default["default"].createElement("div", {
    className: "sendbird-muted-avatar",
    style: {
      height: `${height}px`,
      width: `${width}px`
    }
  }, /*#__PURE__*/React__default["default"].createElement("div", {
    className: "sendbird-muted-avatar__icon"
  }, /*#__PURE__*/React__default["default"].createElement("div", {
    className: "sendbird-muted-avatar__bg",
    style: {
      height: `${height}px`,
      width: `${width}px`
    }
  }), /*#__PURE__*/React__default["default"].createElement(ui_Icon["default"], {
    type: ui_Icon.IconTypes.MUTE,
    fillColor: ui_Icon.IconColors.WHITE,
    width: `${height - 8}px`,
    height: `${width - 8}px`
  })));
}

module.exports = MutedAvatarOverlay;
