// @sendbird/uikit-react/cjs/ui/OpenChannelAvatar.js
'use strict';

var React = require('react');
var ui_Avatar = require('./Avatar.js');
var LocalizationContext = require('../LocalizationContext-df436a16.js');
var utils = require('../utils-58e012fb.js');
require('./ImageRenderer.js');
require('./Icon.js');
require('prop-types');
require('../uuid-2953f4dd.js');
require('../stringSet-d7f78de0.js');
require('../index-0f42efcf.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function ChannelAvatar(_ref) {
  let {
    channel,
    theme,
    height = 56,
    width = 56
  } = _ref;
  const {
    stringSet
  } = React.useContext(LocalizationContext.LocalizationContext);
  const memoizedAvatar = React.useMemo(() => {
    return /*#__PURE__*/React__default["default"].createElement(ui_Avatar["default"], {
      className: "sendbird-chat-header__avatar--open-channel",
      src: utils.getOpenChannelAvatar(channel),
      width: `${width}px`,
      height: `${height}px`,
      alt: (channel === null || channel === void 0 ? void 0 : channel.name) || stringSet.OPEN_CHANNEL_SETTINGS__NO_TITLE
    });
  }, [channel === null || channel === void 0 ? void 0 : channel.coverUrl, theme]);
  return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, memoizedAvatar);
}

module.exports = ChannelAvatar;
