// @sendbird/uikit-react/Channel.js
import React__default from 'react';
import { C as ChannelProvider } from './ChannelProvider-910bbc1f.js';
import ChannelUI from './Channel/components/ChannelUI.js';
import './UserProfileContext-9cfbab4d.js';
import 'prop-types';
import './useSendbirdStateContext.js';
import './withSendbird.js';
import './_rollupPluginBabelHelpers-6bb0305c.js';
import './index-7cdd58b8.js';
import './index-99aabd9e.js';
import './topics-b2b77fb9.js';
import './index-7eb65acf.js';
import './utils/message/getOutgoingMessageState.js';
import './compareIds-defe2bb3.js';
import './const-b3f02313.js';
import '@sendbird/chat/groupChannel';
import './uuid-1adedf62.js';
import '@sendbird/chat/message';
import './consts-95d8566e.js';
import './index-cef1cfff.js';
import './LocalizationContext-076d6d2e.js';
import './stringSet-13bbd4a2.js';
import './ui/Icon.js';
import './index-63949de5.js';
import './ui/Loader.js';
import './ui/ConnectionStatus.js';
import './Channel/components/ChannelHeader.js';
import './utils-30d3ca65.js';
import './ui/IconButton.js';
import './ui/ChannelAvatar.js';
import './ui/Avatar.js';
import './ui/ImageRenderer.js';
import './utils-d86ad6ab.js';
import './MediaQueryContext-11f0e43f.js';
import './utils-4bbbca2c.js';
import './Channel/components/MessageList.js';
import './Channel/components/Message.js';
import './Channel/components/SuggestedMentionList.js';
import './const-221c6a89.js';
import './ThreadProvider-3a4a4db0.js';
import './Thread/context/types.js';
import '@sendbird/chat';
import './ui/DateSeparator.js';
import './color-879cc347.js';
import './ui/MessageInput.js';
import './ui/Button.js';
import 'dompurify';
import './consts-ff014656.js';
import './consts-c41580ae.js';
import './consts-81544e52.js';
import './tokenize-c40f2b57.js';
import './ui/MessageContent.js';
import './ui/UserProfile.js';
import './sendbirdSelectors.js';
import './index-5c1b672d.js';
import './index-306bc4ab.js';
import './index-fe932783.js';
import './ui/MessageItemMenu.js';
import './ui/ContextMenu.js';
import 'react-dom';
import './ui/SortByRow.js';
import './types-4ff4177c.js';
import './ui/MessageItemReactionMenu.js';
import './ui/ReactionButton.js';
import './useLongPress-1303953d.js';
import './ui/EmojiReactions.js';
import './ui/ReactionBadge.js';
import './ui/BottomSheet.js';
import './index-38e61c64.js';
import './ui/UserListItem.js';
import './ui/MutedAvatarOverlay.js';
import './ui/Checkbox.js';
import './ui/Tooltip.js';
import './ui/TooltipWrapper.js';
import './Message/context.js';
import './ui/AdminMessage.js';
import './ui/TextMessageItemBody.js';
import './index-2c3e11ea.js';
import './index.module-54d57ec3.js';
import './ui/MentionLabel.js';
import './ui/LinkLabel.js';
import './ui/FileMessageItemBody.js';
import './ui/TextButton.js';
import './ui/ThumbnailMessageItemBody.js';
import './ui/OGMessageItemBody.js';
import './ui/UnknownMessageItemBody.js';
import './ui/QuoteMessage.js';
import './index-aca30408.js';
import './ui/ThreadReplies.js';
import './ui/VoiceMessageItemBody.js';
import './ui/ProgressBar.js';
import './VoicePlayer/useVoicePlayer.js';
import './index-5c985608.js';
import './VoiceRecorder/context.js';
import './ui/PlaybackTime.js';
import './Channel/components/FileViewer.js';
import './Channel/components/RemoveMessageModal.js';
import './ui/Modal.js';
import './index-9e49ffe3.js';
import './useDirtyGetMentions-aa730e2c.js';
import './Channel/components/UnreadCount.js';
import './Channel/components/FrozenNotification.js';
import './Channel/components/TypingIndicator.js';
import './Channel/components/MessageInput.js';
import './ui/QuoteMessageInput.js';
import './VoiceMessageInputWrapper-dc459df1.js';
import './VoiceRecorder/useVoiceRecorder.js';
import './index-e6e562c5.js';

const Channel = props => {
  return /*#__PURE__*/React__default.createElement(ChannelProvider, {
    channelUrl: props === null || props === void 0 ? void 0 : props.channelUrl,
    isReactionEnabled: props === null || props === void 0 ? void 0 : props.isReactionEnabled,
    isMessageGroupingEnabled: props === null || props === void 0 ? void 0 : props.isMessageGroupingEnabled,
    showSearchIcon: props === null || props === void 0 ? void 0 : props.showSearchIcon,
    animatedMessage: props === null || props === void 0 ? void 0 : props.animatedMessage,
    highlightedMessage: props === null || props === void 0 ? void 0 : props.highlightedMessage,
    startingPoint: props === null || props === void 0 ? void 0 : props.startingPoint,
    onBeforeSendUserMessage: props === null || props === void 0 ? void 0 : props.onBeforeSendUserMessage,
    onBeforeSendFileMessage: props === null || props === void 0 ? void 0 : props.onBeforeSendFileMessage,
    onBeforeUpdateUserMessage: props === null || props === void 0 ? void 0 : props.onBeforeUpdateUserMessage,
    onChatHeaderActionClick: props === null || props === void 0 ? void 0 : props.onChatHeaderActionClick,
    onSearchClick: props === null || props === void 0 ? void 0 : props.onSearchClick,
    onBackClick: props === null || props === void 0 ? void 0 : props.onBackClick,
    replyType: props === null || props === void 0 ? void 0 : props.replyType,
    threadReplySelectType: props === null || props === void 0 ? void 0 : props.threadReplySelectType,
    queries: props === null || props === void 0 ? void 0 : props.queries,
    renderUserProfile: props === null || props === void 0 ? void 0 : props.renderUserProfile,
    filterMessageList: props === null || props === void 0 ? void 0 : props.filterMessageList,
    disableUserProfile: props === null || props === void 0 ? void 0 : props.disableUserProfile,
    disableMarkAsRead: props === null || props === void 0 ? void 0 : props.disableMarkAsRead,
    onReplyInThread: props === null || props === void 0 ? void 0 : props.onReplyInThread,
    onQuoteMessageClick: props === null || props === void 0 ? void 0 : props.onQuoteMessageClick,
    onMessageAnimated: props === null || props === void 0 ? void 0 : props.onMessageAnimated,
    onMessageHighlighted: props === null || props === void 0 ? void 0 : props.onMessageHighlighted
  }, /*#__PURE__*/React__default.createElement(ChannelUI, {
    isLoading: props === null || props === void 0 ? void 0 : props.isLoading,
    renderPlaceholderLoader: props === null || props === void 0 ? void 0 : props.renderPlaceholderLoader,
    renderPlaceholderInvalid: props === null || props === void 0 ? void 0 : props.renderPlaceholderInvalid,
    renderPlaceholderEmpty: props === null || props === void 0 ? void 0 : props.renderPlaceholderEmpty,
    renderChannelHeader: props === null || props === void 0 ? void 0 : props.renderChannelHeader,
    renderMessage: props === null || props === void 0 ? void 0 : props.renderMessage,
    renderMessageInput: props === null || props === void 0 ? void 0 : props.renderMessageInput,
    renderTypingIndicator: props === null || props === void 0 ? void 0 : props.renderTypingIndicator,
    renderCustomSeparator: props === null || props === void 0 ? void 0 : props.renderCustomSeparator,
    renderFileUploadIcon: props === null || props === void 0 ? void 0 : props.renderFileUploadIcon,
    renderVoiceMessageIcon: props === null || props === void 0 ? void 0 : props.renderVoiceMessageIcon,
    renderSendMessageIcon: props === null || props === void 0 ? void 0 : props.renderSendMessageIcon
  }));
};

export { Channel as default };
