// @sendbird/uikit-react/ChannelList.js
import React__default from 'react';
import { C as ChannelListProvider } from './ChannelListProvider-21425002.js';
import ChannelListUI from './ChannelList/components/ChannelListUI.js';
import './_rollupPluginBabelHelpers-6bb0305c.js';
import '@sendbird/chat/groupChannel';
import './topics-b2b77fb9.js';
import './uuid-1adedf62.js';
import './utils-4bbbca2c.js';
import './UserProfileContext-9cfbab4d.js';
import 'prop-types';
import './useSendbirdStateContext.js';
import './withSendbird.js';
import './index-7eb65acf.js';
import './utils/message/getOutgoingMessageState.js';
import './ChannelList/components/ChannelListHeader.js';
import './LocalizationContext-076d6d2e.js';
import './stringSet-13bbd4a2.js';
import './index-99aabd9e.js';
import './index-63949de5.js';
import './ui/Avatar.js';
import './ui/ImageRenderer.js';
import './ui/Icon.js';
import './ChannelList/components/AddChannel.js';
import './ui/IconButton.js';
import './CreateChannel.js';
import './CreateChannel/components/CreateChannelUI.js';
import './CreateChannelProvider-11c8b9a3.js';
import './sendbirdSelectors.js';
import './CreateChannel/components/InviteUsers.js';
import './ui/Modal.js';
import 'react-dom';
import './index-38e61c64.js';
import './ui/Button.js';
import './MediaQueryContext-11f0e43f.js';
import './ui/UserListItem.js';
import './ui/MutedAvatarOverlay.js';
import './ui/Checkbox.js';
import './ui/UserProfile.js';
import './ui/ContextMenu.js';
import './ui/SortByRow.js';
import './CreateChannel/components/SelectChannelType.js';
import './ChannelList/components/ChannelPreview.js';
import './ui/ChannelAvatar.js';
import './utils-d86ad6ab.js';
import './ui/Badge.js';
import './index-5c1b672d.js';
import './index-7cdd58b8.js';
import './ui/Loader.js';
import './index-306bc4ab.js';
import './index-fe932783.js';
import './ui/MentionUserLabel.js';
import './consts-ff014656.js';
import './ui/TextButton.js';
import './color-879cc347.js';
import './Channel/components/TypingIndicator.js';
import './ChannelProvider-910bbc1f.js';
import './compareIds-defe2bb3.js';
import './const-b3f02313.js';
import '@sendbird/chat/message';
import './consts-95d8566e.js';
import './useLongPress-1303953d.js';
import './ChannelList/components/ChannelPreviewAction.js';
import './EditUserProfile.js';
import './index-fb8a5bc3.js';
import './ui/Input.js';
import './actionTypes-ce6caec1.js';
import './index-cef1cfff.js';

const ChannelList = props => {
  return /*#__PURE__*/React__default.createElement(ChannelListProvider, {
    className: props === null || props === void 0 ? void 0 : props.className,
    disableUserProfile: props === null || props === void 0 ? void 0 : props.disableUserProfile,
    allowProfileEdit: props === null || props === void 0 ? void 0 : props.allowProfileEdit,
    onBeforeCreateChannel: props === null || props === void 0 ? void 0 : props.onBeforeCreateChannel,
    onThemeChange: props === null || props === void 0 ? void 0 : props.onThemeChange,
    overrideInviteUser: props === null || props === void 0 ? void 0 : props.overrideInviteUser,
    onProfileEditSuccess: props === null || props === void 0 ? void 0 : props.onProfileEditSuccess,
    onChannelSelect: props === null || props === void 0 ? void 0 : props.onChannelSelect,
    sortChannelList: props === null || props === void 0 ? void 0 : props.sortChannelList,
    queries: props === null || props === void 0 ? void 0 : props.queries,
    disableAutoSelect: props === null || props === void 0 ? void 0 : props.disableAutoSelect,
    activeChannelUrl: props === null || props === void 0 ? void 0 : props.activeChannelUrl,
    isTypingIndicatorEnabled: props === null || props === void 0 ? void 0 : props.isTypingIndicatorEnabled,
    isMessageReceiptStatusEnabled: props === null || props === void 0 ? void 0 : props.isMessageReceiptStatusEnabled
  }, /*#__PURE__*/React__default.createElement(ChannelListUI, {
    renderChannelPreview: props === null || props === void 0 ? void 0 : props.renderChannelPreview,
    renderUserProfile: props === null || props === void 0 ? void 0 : props.renderUserProfile,
    renderHeader: props === null || props === void 0 ? void 0 : props.renderHeader,
    renderPlaceHolderEmptyList: props === null || props === void 0 ? void 0 : props.renderPlaceHolderEmptyList,
    renderPlaceHolderError: props === null || props === void 0 ? void 0 : props.renderPlaceHolderError,
    renderPlaceHolderLoading: props === null || props === void 0 ? void 0 : props.renderPlaceHolderLoading
  }));
};

export { ChannelList as default };
