// @sendbird/uikit-react/CreateOpenChannel.js
import React__default from 'react';
import CreateOpenChannelUI from './CreateOpenChannel/components/CreateOpenChannelUI.js';
import { CreateOpenChannelProvider } from './CreateOpenChannel/context.js';
import './LocalizationContext-076d6d2e.js';
import './stringSet-13bbd4a2.js';
import './index-99aabd9e.js';
import './ui/Avatar.js';
import './ui/ImageRenderer.js';
import './ui/Icon.js';
import 'prop-types';
import './uuid-1adedf62.js';
import './ui/Button.js';
import './index-63949de5.js';
import './ui/Modal.js';
import 'react-dom';
import './index-38e61c64.js';
import './ui/IconButton.js';
import './_rollupPluginBabelHelpers-6bb0305c.js';
import './MediaQueryContext-11f0e43f.js';
import './ui/Input.js';
import './ui/TextButton.js';
import './color-879cc347.js';
import './useSendbirdStateContext.js';
import './withSendbird.js';

function CreateOpenChannel(_ref) {
  let {
    className,
    onCreateChannel,
    onBeforeCreateChannel,
    closeModal,
    renderHeader,
    renderProfileInput
  } = _ref;
  return /*#__PURE__*/React__default.createElement(CreateOpenChannelProvider, {
    className: className,
    onCreateChannel: onCreateChannel,
    onBeforeCreateChannel: onBeforeCreateChannel
  }, /*#__PURE__*/React__default.createElement(CreateOpenChannelUI, {
    closeModal: closeModal,
    renderHeader: renderHeader,
    renderProfileInput: renderProfileInput
  }));
}

export { CreateOpenChannel as default };
