// @sendbird/uikit-react/CreateChannel.js
import React__default from 'react';
import CreateChannel$1 from './CreateChannel/components/CreateChannelUI.js';
import { C as CreateChannelProvider } from './CreateChannelProvider-11c8b9a3.js';
import './CreateChannel/components/InviteUsers.js';
import './_rollupPluginBabelHelpers-6bb0305c.js';
import './LocalizationContext-076d6d2e.js';
import './stringSet-13bbd4a2.js';
import './index-99aabd9e.js';
import './useSendbirdStateContext.js';
import './withSendbird.js';
import './ui/Modal.js';
import 'react-dom';
import './index-38e61c64.js';
import './ui/Button.js';
import './index-63949de5.js';
import 'prop-types';
import './ui/Icon.js';
import './ui/IconButton.js';
import './MediaQueryContext-11f0e43f.js';
import './ui/UserListItem.js';
import './UserProfileContext-9cfbab4d.js';
import './ui/Avatar.js';
import './ui/ImageRenderer.js';
import './uuid-1adedf62.js';
import './ui/MutedAvatarOverlay.js';
import './ui/Checkbox.js';
import './ui/UserProfile.js';
import './sendbirdSelectors.js';
import './topics-b2b77fb9.js';
import './utils-4bbbca2c.js';
import './ui/ContextMenu.js';
import './ui/SortByRow.js';
import './index-7eb65acf.js';
import './utils/message/getOutgoingMessageState.js';
import './CreateChannel/components/SelectChannelType.js';

const CreateChannel = props => {
  const {
    onBeforeCreateChannel,
    userListQuery,
    onCreateChannel,
    overrideInviteUser,
    onCancel,
    renderStepOne
  } = props;
  return /*#__PURE__*/React__default.createElement(CreateChannelProvider, {
    onBeforeCreateChannel: onBeforeCreateChannel,
    userListQuery: userListQuery,
    onCreateChannel: onCreateChannel,
    overrideInviteUser: overrideInviteUser
  }, /*#__PURE__*/React__default.createElement(CreateChannel$1, {
    renderStepOne: renderStepOne,
    onCancel: onCancel
  }));
};

export { CreateChannel as default };
