// @sendbird/uikit-react/OpenChannel.js
import React__default from 'react';
import OpenChannelUI from './OpenChannel/components/OpenChannelUI.js';
import { O as OpenChannelProvider } from './OpenChannelProvider-d3a850e3.js';
import './OpenChannel/components/OpenChannelInput.js';
import './LocalizationContext-076d6d2e.js';
import './stringSet-13bbd4a2.js';
import './index-99aabd9e.js';
import './ui/MessageInput.js';
import './_rollupPluginBabelHelpers-6bb0305c.js';
import 'prop-types';
import './const-221c6a89.js';
import './const-b3f02313.js';
import './ui/IconButton.js';
import './ui/Button.js';
import './index-63949de5.js';
import 'dompurify';
import './consts-ff014656.js';
import './ui/Icon.js';
import './index-7eb65acf.js';
import './utils/message/getOutgoingMessageState.js';
import './consts-c41580ae.js';
import './consts-81544e52.js';
import './tokenize-c40f2b57.js';
import './OpenChannel/components/FrozenChannelNotification.js';
import './OpenChannel/components/OpenChannelHeader.js';
import './ui/Avatar.js';
import './ui/ImageRenderer.js';
import './uuid-1adedf62.js';
import './MediaQueryContext-11f0e43f.js';
import './index-cef1cfff.js';
import './ui/Loader.js';
import './OpenChannel/components/OpenChannelMessageList.js';
import './index-fe932783.js';
import './index-7cdd58b8.js';
import './OpenChannel/components/OpenChannelMessage.js';
import './ui/OpenchannelUserMessage.js';
import './ui/ContextMenu.js';
import 'react-dom';
import './ui/SortByRow.js';
import './ui/UserProfile.js';
import './UserProfileContext-9cfbab4d.js';
import './sendbirdSelectors.js';
import './topics-b2b77fb9.js';
import './utils-4bbbca2c.js';
import './useSendbirdStateContext.js';
import './withSendbird.js';
import './utils-a2009c36.js';
import './index-f7ec60a8.js';
import './useLongPress-1303953d.js';
import './ui/OpenChannelAdminMessage.js';
import './ui/OpenchannelOGMessage.js';
import './ui/LinkLabel.js';
import './index-2c3e11ea.js';
import './index.module-54d57ec3.js';
import './Message/context.js';
import './ui/MentionLabel.js';
import './ui/OpenchannelThumbnailMessage.js';
import './ui/OpenchannelFileMessage.js';
import './ui/TextButton.js';
import './color-879cc347.js';
import './ui/DateSeparator.js';
import './ui/FileViewer.js';
import './index-38e61c64.js';
import './ui/Modal.js';
import './index-9e49ffe3.js';
import './consts-95d8566e.js';
import './compareIds-defe2bb3.js';
import '@sendbird/chat';
import '@sendbird/chat/openChannel';

const OpenChannel = props => {
  return /*#__PURE__*/React__default.createElement(OpenChannelProvider, {
    channelUrl: props === null || props === void 0 ? void 0 : props.channelUrl,
    isMessageGroupingEnabled: props === null || props === void 0 ? void 0 : props.isMessageGroupingEnabled,
    queries: props === null || props === void 0 ? void 0 : props.queries,
    messageLimit: props === null || props === void 0 ? void 0 : props.messageLimit,
    onBeforeSendUserMessage: props === null || props === void 0 ? void 0 : props.onBeforeSendUserMessage,
    onBeforeSendFileMessage: props === null || props === void 0 ? void 0 : props.onBeforeSendFileMessage,
    onChatHeaderActionClick: props === null || props === void 0 ? void 0 : props.onChatHeaderActionClick,
    disableUserProfile: props === null || props === void 0 ? void 0 : props.disableUserProfile,
    renderUserProfile: props === null || props === void 0 ? void 0 : props.renderUserProfile,
    onBackClick: props === null || props === void 0 ? void 0 : props.onBackClick
  }, /*#__PURE__*/React__default.createElement(OpenChannelUI, {
    renderMessage: props === null || props === void 0 ? void 0 : props.renderMessage,
    renderHeader: props === null || props === void 0 ? void 0 : props.renderHeader,
    renderInput: props === null || props === void 0 ? void 0 : props.renderInput,
    renderPlaceHolderEmptyList: props === null || props === void 0 ? void 0 : props.renderPlaceHolderEmptyList,
    renderPlaceHolderError: props === null || props === void 0 ? void 0 : props.renderPlaceHolderError,
    renderPlaceHolderLoading: props === null || props === void 0 ? void 0 : props.renderPlaceHolderLoading
  }));
};

export { OpenChannel as default };
