// @sendbird/uikit-react/OpenChannelList.js
import React__default from 'react';
import OpenChannelListUI from './OpenChannelList/components/OpenChannelListUI.js';
import { O as OpenChannelListProvider } from './OpenChannelListProvider-426cfb39.js';
import './OpenChannelList/components/OpenChannelPreview.js';
import './ui/Avatar.js';
import './ui/ImageRenderer.js';
import './ui/Icon.js';
import 'prop-types';
import './uuid-1adedf62.js';
import './index-63949de5.js';
import './stringSet-13bbd4a2.js';
import './index-cef1cfff.js';
import './LocalizationContext-076d6d2e.js';
import './index-99aabd9e.js';
import './ui/Loader.js';
import './ui/IconButton.js';
import './_rollupPluginBabelHelpers-6bb0305c.js';
import './CreateOpenChannel.js';
import './CreateOpenChannel/components/CreateOpenChannelUI.js';
import './ui/Button.js';
import './ui/Modal.js';
import 'react-dom';
import './index-38e61c64.js';
import './MediaQueryContext-11f0e43f.js';
import './ui/Input.js';
import './ui/TextButton.js';
import './color-879cc347.js';
import './CreateOpenChannel/context.js';
import './useSendbirdStateContext.js';
import './withSendbird.js';
import './consts-95d8566e.js';
import './topics-b2b77fb9.js';

function OpenChannelList(_ref) {
  let {
    // provider
    className,
    queries,
    onChannelSelected,
    // ui
    renderHeader,
    renderChannelPreview,
    renderPlaceHolderEmpty,
    renderPlaceHolderError,
    renderPlaceHolderLoading
  } = _ref;
  return /*#__PURE__*/React__default.createElement(OpenChannelListProvider, {
    className: className,
    queries: queries,
    onChannelSelected: onChannelSelected
  }, /*#__PURE__*/React__default.createElement(OpenChannelListUI, {
    renderHeader: renderHeader,
    renderChannelPreview: renderChannelPreview,
    renderPlaceHolderEmpty: renderPlaceHolderEmpty,
    renderPlaceHolderError: renderPlaceHolderError,
    renderPlaceHolderLoading: renderPlaceHolderLoading
  }));
}

export { OpenChannelList as default };
