// @sendbird/uikit-react/OpenChannelSettings.js
import React__default from 'react';
import OpenChannelUI from './OpenChannelSettings/components/OpenChannelSettingsUI.js';
import { OpenChannelSettingsProvider } from './OpenChannelSettings/context.js';
import './useSendbirdStateContext.js';
import './withSendbird.js';
import './_rollupPluginBabelHelpers-6bb0305c.js';
import './UserProfileContext-9cfbab4d.js';
import 'prop-types';
import './LocalizationContext-076d6d2e.js';
import './stringSet-13bbd4a2.js';
import './index-99aabd9e.js';
import './index-63949de5.js';
import './ui/Icon.js';
import './index-cef1cfff.js';
import './ui/Loader.js';
import './OpenChannelSettings/components/OperatorUI.js';
import './OpenChannelSettings/components/OpenChannelProfile.js';
import './ui/TextButton.js';
import './color-879cc347.js';
import './ui/OpenChannelAvatar.js';
import './ui/Avatar.js';
import './ui/ImageRenderer.js';
import './uuid-1adedf62.js';
import './utils-d86ad6ab.js';
import './OpenChannelSettings/components/EditDetailsModal.js';
import './topics-b2b77fb9.js';
import './ui/Modal.js';
import 'react-dom';
import './index-38e61c64.js';
import './ui/Button.js';
import './ui/IconButton.js';
import './MediaQueryContext-11f0e43f.js';
import './ui/Input.js';
import './index-eaf0d65b.js';
import './ui/ContextMenu.js';
import './ui/SortByRow.js';
import './index-7eb65acf.js';
import './utils/message/getOutgoingMessageState.js';
import './context-63cbe5bd.js';
import './utils-4bbbca2c.js';
import './ui/MutedAvatarOverlay.js';
import './ui/UserProfile.js';
import './sendbirdSelectors.js';
import '@sendbird/chat/openChannel';
import './ui/UserListItem.js';
import './ui/Checkbox.js';
import './ui/Accordion.js';
import './ui/AccordionGroup.js';

const OpenChannelSetting = props => {
  return /*#__PURE__*/React__default.createElement(OpenChannelSettingsProvider, {
    channelUrl: props === null || props === void 0 ? void 0 : props.channelUrl,
    onCloseClick: props === null || props === void 0 ? void 0 : props.onCloseClick,
    onBeforeUpdateChannel: props === null || props === void 0 ? void 0 : props.onBeforeUpdateChannel,
    onChannelModified: props === null || props === void 0 ? void 0 : props.onChannelModified,
    onDeleteChannel: props === null || props === void 0 ? void 0 : props.onDeleteChannel,
    disableUserProfile: props === null || props === void 0 ? void 0 : props.disableUserProfile,
    renderUserProfile: props === null || props === void 0 ? void 0 : props.renderUserProfile
  }, /*#__PURE__*/React__default.createElement(OpenChannelUI, {
    renderOperatorUI: props === null || props === void 0 ? void 0 : props.renderOperatorUI,
    renderParticipantList: props === null || props === void 0 ? void 0 : props.renderParticipantList
  }));
};

export { OpenChannelSetting as default };
