// @sendbird/uikit-react/Thread.js
import React__default from 'react';
import { T as ThreadProvider } from './ThreadProvider-3a4a4db0.js';
import ThreadUI from './Thread/components/ThreadUI.js';
import './index-7cdd58b8.js';
import './index-99aabd9e.js';
import './utils/message/getOutgoingMessageState.js';
import './UserProfileContext-9cfbab4d.js';
import 'prop-types';
import './useSendbirdStateContext.js';
import './withSendbird.js';
import './_rollupPluginBabelHelpers-6bb0305c.js';
import './Thread/context/types.js';
import '@sendbird/chat';
import './topics-b2b77fb9.js';
import '@sendbird/chat/groupChannel';
import './uuid-1adedf62.js';
import '@sendbird/chat/message';
import './consts-95d8566e.js';
import './LocalizationContext-076d6d2e.js';
import './stringSet-13bbd4a2.js';
import './utils-30d3ca65.js';
import './index-63949de5.js';
import './Thread/components/ParentMessageInfo.js';
import './RemoveMessageModal-78e395ad.js';
import './ui/Modal.js';
import 'react-dom';
import './index-38e61c64.js';
import './ui/Button.js';
import './ui/Icon.js';
import './ui/IconButton.js';
import './MediaQueryContext-11f0e43f.js';
import './Thread/components/ParentMessageInfoItem.js';
import './index-7eb65acf.js';
import './ui/ImageRenderer.js';
import './ui/TextButton.js';
import './color-879cc347.js';
import './ui/EmojiReactions.js';
import './ui/ReactionBadge.js';
import './ui/ReactionButton.js';
import './useLongPress-1303953d.js';
import './utils-4bbbca2c.js';
import './ui/ContextMenu.js';
import './ui/SortByRow.js';
import './ui/BottomSheet.js';
import './ui/UserListItem.js';
import './ui/Avatar.js';
import './ui/MutedAvatarOverlay.js';
import './ui/Checkbox.js';
import './ui/UserProfile.js';
import './sendbirdSelectors.js';
import './ui/Tooltip.js';
import './ui/TooltipWrapper.js';
import './Message/context.js';
import './ui/VoiceMessageItemBody.js';
import './ui/ProgressBar.js';
import './VoicePlayer/useVoicePlayer.js';
import './index-5c985608.js';
import './VoiceRecorder/context.js';
import './ui/PlaybackTime.js';
import './ui/Loader.js';
import './index-2c3e11ea.js';
import './index.module-54d57ec3.js';
import './tokenize-c40f2b57.js';
import './ui/MentionLabel.js';
import './ui/LinkLabel.js';
import './ChannelProvider-910bbc1f.js';
import './compareIds-defe2bb3.js';
import './const-b3f02313.js';
import './Channel/components/SuggestedMentionList.js';
import './const-221c6a89.js';
import './ui/FileViewer.js';
import './ui/MessageItemMenu.js';
import './types-4ff4177c.js';
import './ui/MessageItemReactionMenu.js';
import './ui/MessageInput.js';
import 'dompurify';
import './consts-ff014656.js';
import './consts-c41580ae.js';
import './consts-81544e52.js';
import './index-aca30408.js';
import './useDirtyGetMentions-aa730e2c.js';
import './Thread/components/ThreadHeader.js';
import './Thread/components/ThreadList.js';
import './Thread/components/ThreadListItem.js';
import './ui/DateSeparator.js';
import './index-5c1b672d.js';
import './index-306bc4ab.js';
import './index-fe932783.js';
import './ui/TextMessageItemBody.js';
import './ui/OGMessageItemBody.js';
import './ui/FileMessageItemBody.js';
import './ui/ThumbnailMessageItemBody.js';
import './ui/UnknownMessageItemBody.js';
import 'date-fns';
import './Thread/components/ThreadMessageInput.js';
import './VoiceMessageInputWrapper-dc459df1.js';
import './VoiceRecorder/useVoiceRecorder.js';
import './index-e6e562c5.js';
import './index-cef1cfff.js';

const Thread = props => {
  const {
    // props
    className,
    // ThreadProviderProps
    channelUrl,
    message,
    onHeaderActionClick,
    onMoveToParentMessage,
    // ThreadUIProps
    renderHeader,
    renderParentMessageInfo,
    renderMessage,
    renderMessageInput,
    renderCustomSeparator,
    renderParentMessageInfoPlaceholder,
    renderThreadListPlaceHolder,
    renderFileUploadIcon,
    renderVoiceMessageIcon,
    renderSendMessageIcon
  } = props;
  return /*#__PURE__*/React__default.createElement("div", {
    className: `sendbird-thread ${className}`
  }, /*#__PURE__*/React__default.createElement(ThreadProvider, {
    channelUrl: channelUrl,
    message: message,
    onHeaderActionClick: onHeaderActionClick,
    onMoveToParentMessage: onMoveToParentMessage
  }, /*#__PURE__*/React__default.createElement(ThreadUI, {
    renderHeader: renderHeader,
    renderParentMessageInfo: renderParentMessageInfo,
    renderMessage: renderMessage,
    renderMessageInput: renderMessageInput,
    renderCustomSeparator: renderCustomSeparator,
    renderParentMessageInfoPlaceholder: renderParentMessageInfoPlaceholder,
    renderThreadListPlaceHolder: renderThreadListPlaceHolder,
    renderFileUploadIcon: renderFileUploadIcon,
    renderVoiceMessageIcon: renderVoiceMessageIcon,
    renderSendMessageIcon: renderSendMessageIcon
  })));
};

export { Thread as default };
