// @sendbird/uikit-react/Channel/components/RemoveMessageModal.js
import React__default, { useContext } from 'react';
import Modal from '../../ui/Modal.js';
import { ButtonTypes } from '../../ui/Button.js';
import { a as LocalizationContext } from '../../LocalizationContext-076d6d2e.js';
import { u as useChannelContext } from '../../ChannelProvider-910bbc1f.js';
import 'react-dom';
import '../../index-38e61c64.js';
import '../../ui/Icon.js';
import 'prop-types';
import '../../ui/IconButton.js';
import '../../_rollupPluginBabelHelpers-6bb0305c.js';
import '../../index-63949de5.js';
import '../../stringSet-13bbd4a2.js';
import '../../MediaQueryContext-11f0e43f.js';
import '../../index-99aabd9e.js';
import '../../UserProfileContext-9cfbab4d.js';
import '../../useSendbirdStateContext.js';
import '../../withSendbird.js';
import '../../index-7cdd58b8.js';
import '../../topics-b2b77fb9.js';
import '../../index-7eb65acf.js';
import '../../utils/message/getOutgoingMessageState.js';
import '../../compareIds-defe2bb3.js';
import '../../const-b3f02313.js';
import '@sendbird/chat/groupChannel';
import '../../uuid-1adedf62.js';
import '@sendbird/chat/message';
import '../../consts-95d8566e.js';

const RemoveMessage = props => {
  var _message$threadInfo;
  const {
    onSubmit = () => {/* noop */},
    onCancel,
    message
  } = props;
  const {
    stringSet
  } = useContext(LocalizationContext);
  const {
    deleteMessage
  } = useChannelContext();
  return /*#__PURE__*/React__default.createElement(Modal, {
    type: ButtonTypes.DANGER,
    disabled: (message === null || message === void 0 ? void 0 : (_message$threadInfo = message.threadInfo) === null || _message$threadInfo === void 0 ? void 0 : _message$threadInfo.replyCount) > 0,
    onCancel: onCancel,
    onSubmit: () => {
      deleteMessage(message).then(() => {
        onSubmit();
        onCancel();
      });
    },
    submitText: stringSet.MESSAGE_MENU__DELETE,
    titleText: stringSet.MODAL__DELETE_MESSAGE__TITLE
  });
};

export { RemoveMessage as default };
