// @sendbird/uikit-react/ChannelList/components/AddChannel.js
import React__default, { useState } from 'react';
import IconButton from '../../ui/IconButton.js';
import Icon, { IconTypes, IconColors } from '../../ui/Icon.js';
import CreateChannel from '../../CreateChannel.js';
import useSendbirdStateContext from '../../useSendbirdStateContext.js';
import { u as useChannelListContext } from '../../ChannelListProvider-21425002.js';
import '../../_rollupPluginBabelHelpers-6bb0305c.js';
import 'prop-types';
import '../../CreateChannel/components/CreateChannelUI.js';
import '../../CreateChannelProvider-11c8b9a3.js';
import '../../sendbirdSelectors.js';
import '../../topics-b2b77fb9.js';
import '../../utils-4bbbca2c.js';
import '../../CreateChannel/components/InviteUsers.js';
import '../../LocalizationContext-076d6d2e.js';
import '../../stringSet-13bbd4a2.js';
import '../../index-99aabd9e.js';
import '../../ui/Modal.js';
import 'react-dom';
import '../../index-38e61c64.js';
import '../../ui/Button.js';
import '../../index-63949de5.js';
import '../../MediaQueryContext-11f0e43f.js';
import '../../ui/UserListItem.js';
import '../../UserProfileContext-9cfbab4d.js';
import '../../ui/Avatar.js';
import '../../ui/ImageRenderer.js';
import '../../uuid-1adedf62.js';
import '../../ui/MutedAvatarOverlay.js';
import '../../ui/Checkbox.js';
import '../../ui/UserProfile.js';
import '../../ui/ContextMenu.js';
import '../../ui/SortByRow.js';
import '../../index-7eb65acf.js';
import '../../utils/message/getOutgoingMessageState.js';
import '../../CreateChannel/components/SelectChannelType.js';
import '../../withSendbird.js';
import '@sendbird/chat/groupChannel';

const AddChannel = () => {
  var _state$config;
  const [showModal, setShowModal] = useState(false);
  const state = useSendbirdStateContext();
  const isOnline = state === null || state === void 0 ? void 0 : (_state$config = state.config) === null || _state$config === void 0 ? void 0 : _state$config.isOnline;
  const disabled = !isOnline;
  const {
    overrideInviteUser,
    onBeforeCreateChannel
  } = useChannelListContext();
  return /*#__PURE__*/React__default.createElement(React__default.Fragment, null, /*#__PURE__*/React__default.createElement(IconButton, {
    height: "32px",
    width: "32px",
    onClick: () => {
      setShowModal(true);
    },
    disabled: disabled
  }, /*#__PURE__*/React__default.createElement(Icon, {
    type: IconTypes.CREATE,
    fillColor: IconColors.PRIMARY,
    width: "24px",
    height: "24px"
  })), showModal && /*#__PURE__*/React__default.createElement(CreateChannel, {
    onCancel: () => {
      setShowModal(false);
    },
    overrideInviteUser: overrideInviteUser,
    onCreateChannel: () => {
      setShowModal(false);
    },
    onBeforeCreateChannel: onBeforeCreateChannel
  }));
};

export { AddChannel, AddChannel as default };
